### PR TITLE
New ownership concept

### PR DESCRIPTION
**Pull request by IUdalov**
This is attempt to split up logic, simplify our code.
Previous PR https://github.com/effil/effil/pull/94.

### CODE REVIEW

File: src/cpp/gc-data.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+// Derived classes always managed by corresponding views.
+class GCData {
+public:
+    GCData() = default;
+    virtual ~GCData() = default;
+
+    // List of weak references to nested objects
+    std::unordered_set<GCHandle> refers() const;
+
+public://protected:
```
Comment (mihacooper) [inline src/cpp/gc-data.h]: trash

File: src/cpp/gc-data.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    void addReference(GCHandle handle);
+    void removeReference(GCHandle handle);
+
+public:
+    GCData(const GCData&) = delete;
+    GCData& operator=(const GCData&) = delete;
+
+private:
+    mutable SpinMutex mutex_;
+    std::unordered_multiset<GCHandle> weakRefs_;
```
Comment (mihacooper) [inline src/cpp/gc-data.h:28]: why it's multiset?

File: CMakeLists.txt [hunk 3]
```diff
@@ -40,7 +40,7 @@ include_directories(src/cpp libs/sol/single/sol ${LUA_INCLUDE_DIR})
 FILE(GLOB SOURCES src/cpp/*.cpp src/cpp/*.h)
 FILE(GLOB LUA_SOURCES src/lua/*.lua)
 
-add_library(effil SHARED ${SOURCES})
+add_library(effil SHARED ${SOURCES} src/cpp/gc-object.h)
```
Comment (mihacooper) [inline CMakeLists.txt]: You forgot to remove this again

File: src/cpp/gc-data.h (hunk 2 above)
Comment (IUdalov) [reply src/cpp/gc-data.h:28]: Multiset helps us store several week refs to the same object, thus following code works:
```
test.gc.store_same_value = function()
    local fill = function (c)
        local a = effil.table {}
        c:push(a)
        c:push(a)
    end

    local c = effil.channel()
    fill(c)

    c:pop()
    collectgarbage()
    effil.gc.collect()
    c:pop()[1] = 0
end
```

File: src/cpp/gc-data.h (hunk 1 above)
Comment (IUdalov) [reply src/cpp/gc-data.h]: Fixed

File: CMakeLists.txt (hunk 3 above)
Comment (IUdalov) [reply CMakeLists.txt]: Fixed

**Pull request merged**